File: web-frontend/modules/core/formula/parser/formulaValidationVisitor.js
import BaserowFormulaVisitor from '@baserow/modules/core/formula/parser/generated/BaserowFormulaVisitor'
import { InvalidFormulaType } from '@baserow/modules/core/formula/parser/errors.js'

/**
 * A visitor that validates formula functions and their arguments during parsing.
 * Each function can define custom validation logic via validateArguments().
 *
 * This is used to validate formulas before execution, catching errors early
 * and providing better user feedback.
 */
export default class BaserowFormulaValidationVisitor extends BaserowFormulaVisitor {
  /**
   * @param {FunctionCollection} functions - The collection of available formula functions
   * @param {Object} validationContext - Context needed for validation (e.g., dataProviderRegistry)
   */
  constructor(functions, validationContext = {}) {
    super()
    this.functions = functions
    this.validationContext = validationContext
  }

  visitRoot(ctx) {
    return ctx.expr().accept(this)
  }

  visitFieldReference(ctx) {
    throw new InvalidFormulaType("Unsupported function 'field'.")
  }

  visitStringLiteral(ctx) {
    return this.processString(ctx)
  }

  visitDecimalLiteral(ctx) {
    return parseFloat(ctx.getText())
  }

  visitBooleanLiteral(ctx) {
    return ctx.TRUE() !== null
  }

  visitBrackets(ctx) {
    return ctx.expr().accept(this)
  }

  visitIdentifier(ctx) {
    return ctx.getText()
  }

  visitIntegerLiteral(ctx) {
    return parseInt(ctx.getText())
  }

  visitLeftWhitespaceOrComments(ctx) {
    return ctx.expr().accept(this)
  }

  visitRightWhitespaceOrComments(ctx) {
    return ctx.expr().accept(this)
  }

  processString(ctx) {
    const literalWithoutOuterQuotes = ctx.getText().slice(1, -1)
    let literal
    if (ctx.SINGLEQ_STRING_LITERAL() !== null) {
      literal = literalWithoutOuterQuotes.replace(/\\'/g, "'")
    } else {
      literal = literalWithoutOuterQuotes.replace(/\\"/g, '"')
    }
    return literal
  }

  /**
   * Visit a function call and validate its arguments.
   */
  visitFunctionCall(ctx) {
    const functionName = ctx.func_name().getText().toLowerCase()
    const functionArgumentExpressions = ctx.expr()
    let formulaFunctionType = null
    try {
      formulaFunctionType = this.functions.get(functionName)
    } catch(e) {
      throw new InvalidFormulaType(`Unsupported function '${functionName}'.`)
    }

    // Accept the argument expressions, then before parsing them,
    // confirm that we have the valid number of arguments.
    const acceptedArgs = Array.from(functionArgumentExpressions, (expr) =>
      expr.accept(this)
    )
    formulaFunctionType.validateNumberOfArgs(acceptedArgs, true)

    // Now that we have checked we have the correct number of args,
    // we can safely validate their types and values.
    const argsParsed = formulaFunctionType.parseArgs(acceptedArgs)
    formulaFunctionType.validateArgs(argsParsed, this.validationContext, ctx)

    // Continue visiting children to validate nested function calls
    for (const expr of functionArgumentExpressions) {
      expr.accept(this)
    }

    return null
  }
}

File: backend/src/baserow/core/formula/parser/formula_validation_visitor.py
from typing import TYPE_CHECKING, Optional

from baserow.core.formula.exceptions import InvalidRuntimeFormula
from baserow.core.formula.parser.exceptions import (
    FieldByIdReferencesAreDeprecated,
    FormulaFunctionTypeDoesNotExist,
    InvalidNumberOfArguments,
)
from baserow.core.formula.parser.generated.BaserowFormula import BaserowFormula
from baserow.core.formula.parser.generated.BaserowFormulaVisitor import (
    BaserowFormulaVisitor,
)

if TYPE_CHECKING:
    from baserow.core.formula import FunctionCollection
    from baserow.core.formula.registries import DataProviderTypeRegistry


class BaserowFormulaValidationVisitor(BaserowFormulaVisitor):
    """
    A Baserow formula visitor which is responsible for validating a formula's
    function and its arguments.
    """

    def __init__(
        self,
        functions: "FunctionCollection",
        data_provider_type_registry: Optional["DataProviderTypeRegistry"] = None,
    ):
        self.functions = functions
        self.data_provider_type_registry = data_provider_type_registry

    def visitRoot(self, ctx: BaserowFormula.RootContext):
        return ctx.expr().accept(self)

    def visitStringLiteral(self, ctx: BaserowFormula.StringLiteralContext):
        # noinspection PyTypeChecker
        return self.process_string(ctx)

    def process_string(self, ctx):
        literal_without_outer_quotes = ctx.getText()[1:-1]
        if ctx.SINGLEQ_STRING_LITERAL() is not None:
            literal = literal_without_outer_quotes.replace("\\'", "'")
        else:
            literal = literal_without_outer_quotes.replace('\\"', '"')
        return literal

    def visitDecimalLiteral(self, ctx: BaserowFormula.DecimalLiteralContext):
        return float(ctx.getText())

    def visitBooleanLiteral(self, ctx: BaserowFormula.BooleanLiteralContext):
        return ctx.TRUE() is not None

    def visitBrackets(self, ctx: BaserowFormula.BracketsContext):
        return ctx.expr().accept(self)

    def visitIdentifier(self, ctx: BaserowFormula.IdentifierContext):
        return ctx.getText()

    def visitIntegerLiteral(self, ctx: BaserowFormula.IntegerLiteralContext):
        return int(ctx.getText())

    def visitFieldByIdReference(self, ctx: BaserowFormula.FieldByIdReferenceContext):
        raise FieldByIdReferencesAreDeprecated()

    def visitLeftWhitespaceOrComments(
        self, ctx: BaserowFormula.LeftWhitespaceOrCommentsContext
    ):
        return ctx.expr().accept(self)

    def visitRightWhitespaceOrComments(
        self, ctx: BaserowFormula.RightWhitespaceOrCommentsContext
    ):
        return ctx.expr().accept(self)

    def visitFieldReference(self, ctx: BaserowFormula.FieldReferenceContext):
        """
        Handle field('name') syntax. There is no native support for this function
        in non-database formulas, so we raise an error.
        """

        raise InvalidRuntimeFormula("'field' is not a a supported function")

    def visitFunctionCall(self, ctx: BaserowFormula.FunctionCallContext):
        """
        Visits a function call node in the parse tree. For each function we encounter,
        we validate its args using the corresponding function type's `validate_args`
        method.

        :param ctx: The function call context from the parse tree.
        :raises InvalidNumberOfArguments: If the number of arguments provided to the
            function does not match the expected number.
        :return: The result of visiting child nodes.
        """

        accepted_args = [expr.accept(self) for expr in ctx.expr()]
        function_name = ctx.func_name().getText().lower()
        try:
            formula_function_type = self.functions.get(function_name)
        except FormulaFunctionTypeDoesNotExist:
            raise InvalidRuntimeFormula(f"Unsupported function '{function_name}'.")
        if not formula_function_type.validate_number_of_args(accepted_args):
            raise InvalidNumberOfArguments(formula_function_type, len(accepted_args))
        args_parsed = formula_function_type.parse_args(accepted_args)
        formula_function_type.validate_args(
            args_parsed,
            validation_context={
                "data_provider_type_registry": self.data_provider_type_registry
            },
        )
